Add modifier to the inaccessible classes

From investigation https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1599:

```
[ERROR] Linkage Checker rule found 1 error. Linkage error report:
Class com.sun.xml.internal.ws.api.BindingID$SOAPHTTPImpl is not accessible;
  referenced by 1 class file
    com.sun.tools.internal.ws.wscompile.WsgenOptions (com.sun:tools:1.8)
  Cause:
    Unknown
```

We want to know the accessor of `com.sun.xml.internal.ws.api.BindingID$SOAPHTTPImpl ` class. Public, protected, private, or default?

So should the message be

```
[ERROR] Linkage Checker rule found 1 error. Linkage error report:
Class com.sun.xml.internal.ws.api.BindingID$SOAPHTTPImpl is private;
  referenced by 1 class file
    com.sun.tools.internal.ws.wscompile.WsgenOptions (com.sun:tools:1.8)
  Cause:
    Unknown
```

and

```
[ERROR] Linkage Checker rule found 1 error. Linkage error report:
Class com.sun.xml.internal.ws.api.BindingID$SOAPHTTPImpl is protected;
  referenced by 1 class file that is not a subclass:
    com.sun.tools.internal.ws.wscompile.WsgenOptions (com.sun:tools:1.8)
  Cause:
    Unknown
```

and


```
[ERROR] Linkage Checker rule found 1 error. Linkage error report:
Class com.sun.xml.internal.ws.api.BindingID$SOAPHTTPImpl has default access;
  and is referenced by 1 class file in a different package:
    com.sun.tools.internal.ws.wscompile.WsgenOptions (com.sun:tools:1.8)
  Cause:
    Unknown
```

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/AccessModifier.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/AccessModifier.java
@@ -20,34 +20,30 @@ import org.apache.bcel.Const;
 
 /** A modifier that controls access to classes, fields, or methods. */
 enum AccessModifier {
-  publicAccess,
-  privateAccess,
-  protectedAccess,
-  defaultAccess;
+  PUBLIC(" is public"),
+  PRIVATE(" is private"),
+  PROTECTED(" is private"),
+  DEFAULT(" has default access");
+
+  private String description;
+
+  AccessModifier(String description) {
+    this.description = description;
+  }
 
   static AccessModifier fromFlag(int modifierFlag) {
     if ((modifierFlag & Const.ACC_PUBLIC) != 0) {
-      return publicAccess;
+      return PUBLIC;
     } else if ((modifierFlag & Const.ACC_PRIVATE) != 0) {
-      return privateAccess;
+      return PRIVATE;
     } else if ((modifierFlag & Const.ACC_PROTECTED) != 0) {
-      return protectedAccess;
+      return PROTECTED;
     } else {
-      return defaultAccess;
+      return DEFAULT;
     }
   }
 
   String describe(String item) {
-    switch (this) {
-      case publicAccess:
-        return item + " is public";
-      case privateAccess:
-        return item + " is private";
-      case protectedAccess:
-        return item + " is protected";
-      case defaultAccess:
-        return item + " has default access";
-    }
-    return item + " has unknown access";
+    return item + description;
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/AccessModifier.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/AccessModifier.java
@@ -20,10 +20,10 @@ import org.apache.bcel.Const;
 
 /** A modifier that controls access to classes, fields, or methods. */
 enum AccessModifier {
-  PUBLIC(" is public"),
-  PRIVATE(" is private"),
-  PROTECTED(" is private"),
-  DEFAULT(" has default access");
+  PUBLIC("is public"),
+  PRIVATE("is private"),
+  PROTECTED("is private"),
+  DEFAULT("has default access");
 
   private String description;
 
@@ -44,6 +44,6 @@ enum AccessModifier {
   }
 
   String describe(String item) {
-    return item + description;
+    return item + " " + description;
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/AccessModifier.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/AccessModifier.java
@@ -1,0 +1,53 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import org.apache.bcel.Const;
+
+/** A modifier that controls access to classes, fields, or methods. */
+enum AccessModifier {
+  publicAccess,
+  privateAccess,
+  protectedAccess,
+  defaultAccess;
+
+  static AccessModifier fromFlag(int modifierFlag) {
+    if ((modifierFlag & Const.ACC_PUBLIC) != 0) {
+      return publicAccess;
+    } else if ((modifierFlag & Const.ACC_PRIVATE) != 0) {
+      return privateAccess;
+    } else if ((modifierFlag & Const.ACC_PROTECTED) != 0) {
+      return protectedAccess;
+    } else {
+      return defaultAccess;
+    }
+  }
+
+  String describe(String item) {
+    switch (this) {
+      case publicAccess:
+        return item + " is public";
+      case privateAccess:
+        return item + " is private";
+      case protectedAccess:
+        return item + " is protected";
+      case defaultAccess:
+        return item + " has default access";
+    }
+    return item + " has unknown access";
+  }
+}

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
@@ -17,15 +17,54 @@
 package com.google.cloud.tools.opensource.classpath;
 
 /**
- * The {@code classSymbol} is inaccessible to the {@code sourceClass} as per {@code sourceClass}'s
- * definition of the class symbol.
+ * The {@code classSymbol} with {@code modifier} is inaccessible to the {@code sourceClass} as per
+ * {@code sourceClass}'s definition of the class symbol.
  *
  * <p>If the source class is in a different package than the target class, the class or one of its
  * enclosing types is not public. If the source class is in the same package, the class or one of
  * its enclosing types is private.
  */
 final class InaccessibleClassProblem extends LinkageProblem {
-  InaccessibleClassProblem(ClassFile sourceClass, ClassFile targetClass, Symbol classSymbol) {
+  private AccessModifier modifier;
+
+  InaccessibleClassProblem(
+      ClassFile sourceClass, ClassFile targetClass, Symbol classSymbol, AccessModifier modifier) {
     super("is not accessible", sourceClass, classSymbol, targetClass);
+    this.modifier = modifier;
+  }
+
+  @Override
+  public final String toString() {
+    StringBuilder message = new StringBuilder();
+    message.append("Class " + getTargetClass().getBinaryName());
+    switch (modifier) {
+      case publicAccess:
+        message.append(" is public");
+        break;
+      case privateAccess:
+        message.append(" is private");
+        break;
+      case defaultAccess:
+        message.append(" has default access");
+    }
+
+    message.append(" and referenced by " + getSourceClass().getBinaryName());
+    if (modifier == AccessModifier.defaultAccess) {
+      message.append("(different package)");
+    }
+
+    return message.toString();
+  }
+
+  @Override
+  public String formatSymbolProblem() {
+    String result = modifier.describe(getSymbol().toString());
+    ClassFile targetClass = getTargetClass();
+    if (targetClass != null) {
+      String jarInfo = "(" + targetClass.getClassPathEntry() + ") ";
+      result = jarInfo + result;
+    }
+
+    return result;
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
@@ -73,11 +73,8 @@ final class InaccessibleClassProblem extends LinkageProblem {
     if (modifier == AccessModifier.defaultAccess) {
       return String.format(
           "%s;\n  referenced by %d class file%s in a different package\n",
-          formatSymbolProblem(),
-          referenceCount,
-          referenceCount > 1 ? "s" : "");
+          formatSymbolProblem(), referenceCount, referenceCount > 1 ? "s" : "");
     }
-      return super.formatSymbolProblemWithReferenceCount(referenceCount);
-
+    return super.formatSymbolProblemWithReferenceCount(referenceCount);
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
@@ -38,18 +38,18 @@ final class InaccessibleClassProblem extends LinkageProblem {
     StringBuilder message = new StringBuilder();
     message.append("Class " + getTargetClass().getBinaryName());
     switch (modifier) {
-      case publicAccess:
+      case PUBLIC:
         message.append(" is public");
         break;
-      case privateAccess:
+      case PRIVATE:
         message.append(" is private");
         break;
-      case defaultAccess:
+      case DEFAULT:
         message.append(" has default access");
     }
 
     message.append(" and referenced by " + getSourceClass().getBinaryName());
-    if (modifier == AccessModifier.defaultAccess) {
+    if (modifier == AccessModifier.DEFAULT) {
       message.append("(different package)");
     }
 
@@ -70,7 +70,7 @@ final class InaccessibleClassProblem extends LinkageProblem {
 
   @Override
   protected String formatSymbolProblemWithReferenceCount(int referenceCount) {
-    if (modifier == AccessModifier.defaultAccess) {
+    if (modifier == AccessModifier.DEFAULT) {
       return String.format(
           "%s;\n  referenced by %d class file%s in a different package\n",
           formatSymbolProblem(), referenceCount, referenceCount > 1 ? "s" : "");

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleClassProblem.java
@@ -67,4 +67,17 @@ final class InaccessibleClassProblem extends LinkageProblem {
 
     return result;
   }
+
+  @Override
+  protected String formatSymbolProblemWithReferenceCount(int referenceCount) {
+    if (modifier == AccessModifier.defaultAccess) {
+      return String.format(
+          "%s;\n  referenced by %d class file%s in a different package\n",
+          formatSymbolProblem(),
+          referenceCount,
+          referenceCount > 1 ? "s" : "");
+    }
+      return super.formatSymbolProblemWithReferenceCount(referenceCount);
+
+  }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleMemberProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/InaccessibleMemberProblem.java
@@ -25,7 +25,11 @@ package com.google.cloud.tools.opensource.classpath;
  * target class, the member is not protected or public.
  */
 final class InaccessibleMemberProblem extends LinkageProblem {
-  InaccessibleMemberProblem(ClassFile sourceClass, ClassFile targetClass, Symbol symbol) {
+  private AccessModifier modifier;
+
+  InaccessibleMemberProblem(
+      ClassFile sourceClass, ClassFile targetClass, Symbol symbol, AccessModifier modifier) {
     super("is not accessible", sourceClass, symbol, targetClass);
+    this.modifier = modifier;
   }
 }

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java
@@ -251,7 +251,9 @@ public class LinkageChecker {
           classPathEntry == null ? null : new ClassFile(classPathEntry, targetClassName);
 
       if (!isClassAccessibleFrom(targetJavaClass, sourceClassName)) {
-        return Optional.of(new InaccessibleClassProblem(sourceClassFile, targetClassFile, symbol));
+        AccessModifier modifier = AccessModifier.fromFlag(targetJavaClass.getModifiers());
+        return Optional.of(
+            new InaccessibleClassProblem(sourceClassFile, targetClassFile, symbol, modifier));
       }
 
       if (targetJavaClass.isInterface() != symbol.isInterfaceMethod()) {
@@ -284,8 +286,10 @@ public class LinkageChecker {
             String actualMethodDescriptor = method.getSignature();
             if (actualMethodDescriptor.equals(expectedMethodDescriptor)) {
               if (!isMemberAccessibleFrom(javaClass, method, sourceClassName)) {
+                AccessModifier modifier = AccessModifier.fromFlag(method.getModifiers());
                 return Optional.of(
-                    new InaccessibleMemberProblem(sourceClassFile, targetClassFile, symbol));
+                    new InaccessibleMemberProblem(
+                        sourceClassFile, targetClassFile, symbol, modifier));
               }
               // The method is found and accessible. Returning no error.
               return Optional.empty();
@@ -417,15 +421,19 @@ public class LinkageChecker {
           classFileLocation == null ? null : new ClassFile(classFileLocation, targetClassName);
 
       if (!isClassAccessibleFrom(targetJavaClass, sourceClassName)) {
-        return Optional.of(new InaccessibleClassProblem(sourceClassFile, targetClassFile, symbol));
+        AccessModifier modifier = AccessModifier.fromFlag(targetJavaClass.getModifiers());
+        return Optional.of(
+            new InaccessibleClassProblem(sourceClassFile, targetClassFile, symbol, modifier));
       }
 
       for (JavaClass javaClass : getClassHierarchy(targetJavaClass)) {
         for (Field field : javaClass.getFields()) {
           if (field.getName().equals(fieldName)) {
             if (!isMemberAccessibleFrom(javaClass, field, sourceClassName)) {
+              AccessModifier modifier = AccessModifier.fromFlag(field.getModifiers());
               return Optional.of(
-                  new InaccessibleMemberProblem(sourceClassFile, targetClassFile, symbol));
+                  new InaccessibleMemberProblem(
+                      sourceClassFile, targetClassFile, symbol, modifier));
             }
             // The field is found and accessible. Returning no error.
             return Optional.empty();
@@ -513,7 +521,9 @@ public class LinkageChecker {
 
       if (!isClassAccessibleFrom(targetClass, sourceClassName)
           && classDumper.isClassSymbolReferenceUsed(sourceClassName, symbol)) {
-        return Optional.of(new InaccessibleClassProblem(sourceClassFile, targetClassFile, symbol));
+        AccessModifier modifier = AccessModifier.fromFlag(targetClass.getModifiers());
+        return Optional.of(
+            new InaccessibleClassProblem(sourceClassFile, targetClassFile, symbol, modifier));
       }
       return Optional.empty();
     } catch (ClassNotFoundException ex) {

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java
@@ -138,6 +138,14 @@ public abstract class LinkageProblem {
     return result;
   }
 
+  protected String formatSymbolProblemWithReferenceCount(int referenceCount) {
+    return String.format(
+        "%s;\n  referenced by %d class file%s\n",
+        this.formatSymbolProblem(),
+        referenceCount,
+        referenceCount > 1 ? "s" : "");
+  }
+
   /** Returns mapping from symbol problem description to the names of the source classes. */
   public static ImmutableMap<String, ImmutableSet<String>> groupBySymbolProblem(
       Iterable<LinkageProblem> linkageProblems) {
@@ -185,11 +193,7 @@ public abstract class LinkageProblem {
               LinkageProblem firstProblem = Iterables.getFirst(problems, null);
               int referenceCount = problems.size();
               output.append(
-                  String.format(
-                      "%s;\n  referenced by %d class file%s\n",
-                      firstProblem.formatSymbolProblem(),
-                      referenceCount,
-                      referenceCount > 1 ? "s" : ""));
+                  firstProblem.formatSymbolProblemWithReferenceCount(referenceCount));
               ImmutableSet.Builder<LinkageProblemCause> causesBuilder = ImmutableSet.builder();
               problems.forEach(
                   problem -> {

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblem.java
@@ -141,9 +141,7 @@ public abstract class LinkageProblem {
   protected String formatSymbolProblemWithReferenceCount(int referenceCount) {
     return String.format(
         "%s;\n  referenced by %d class file%s\n",
-        this.formatSymbolProblem(),
-        referenceCount,
-        referenceCount > 1 ? "s" : "");
+        this.formatSymbolProblem(), referenceCount, referenceCount > 1 ? "s" : "");
   }
 
   /** Returns mapping from symbol problem description to the names of the source classes. */
@@ -192,8 +190,7 @@ public abstract class LinkageProblem {
               // problems all have the same symbol problem
               LinkageProblem firstProblem = Iterables.getFirst(problems, null);
               int referenceCount = problems.size();
-              output.append(
-                  firstProblem.formatSymbolProblemWithReferenceCount(referenceCount));
+              output.append(firstProblem.formatSymbolProblemWithReferenceCount(referenceCount));
               ImmutableSet.Builder<LinkageProblemCause> causesBuilder = ImmutableSet.builder();
               problems.forEach(
                   problem -> {

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ExclusionFilesTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ExclusionFilesTest.java
@@ -111,7 +111,8 @@ public class ExclusionFilesTest {
         new InaccessibleMemberProblem(
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
-            new FieldSymbol("com.google.Foo", "fieldA", "Ljava.lang.String;"));
+            new FieldSymbol("com.google.Foo", "fieldA", "Ljava.lang.String;"),
+            AccessModifier.privateAccess);
     boolean result = matchers.get(0).match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -129,7 +130,8 @@ public class ExclusionFilesTest {
         new InaccessibleMemberProblem(
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
-            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false));
+            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
+            AccessModifier.privateAccess);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -147,7 +149,8 @@ public class ExclusionFilesTest {
         new InaccessibleMemberProblem(
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
-            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false));
+            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
+            AccessModifier.privateAccess);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -166,7 +169,8 @@ public class ExclusionFilesTest {
         new InaccessibleMemberProblem(
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.cloud.google.Foo"),
-            new MethodSymbol("com.google.cloud.Foo", "methodA", "()Ljava.lang.String;", false));
+            new MethodSymbol("com.google.cloud.Foo", "methodA", "()Ljava.lang.String;", false),
+            AccessModifier.privateAccess);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -185,7 +189,8 @@ public class ExclusionFilesTest {
         new InaccessibleMemberProblem(
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.googler.Foo"),
-            new MethodSymbol("com.googler.Foo", "methodA", "()Ljava.lang.String;", false));
+            new MethodSymbol("com.googler.Foo", "methodA", "()Ljava.lang.String;", false),
+            AccessModifier.privateAccess);
     boolean result = matcher.match(linkageProblemToMatch);
     assertFalse(result);
   }
@@ -204,7 +209,8 @@ public class ExclusionFilesTest {
             new ClassFile(
                 new ClassPathEntry(Paths.get("dummy.jar")), "reactor.core.publisher.Traces"),
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
-            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false));
+            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
+            AccessModifier.privateAccess);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -222,7 +228,8 @@ public class ExclusionFilesTest {
         new InaccessibleMemberProblem(
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Bar"),
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
-            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false));
+            new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
+            AccessModifier.privateAccess);
     boolean result = matcher.match(linkageProblemToMatch); // No match
     assertFalse(result);
   }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ExclusionFilesTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ExclusionFilesTest.java
@@ -112,7 +112,7 @@ public class ExclusionFilesTest {
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
             new FieldSymbol("com.google.Foo", "fieldA", "Ljava.lang.String;"),
-            AccessModifier.privateAccess);
+            AccessModifier.PRIVATE);
     boolean result = matchers.get(0).match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -131,7 +131,7 @@ public class ExclusionFilesTest {
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
             new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
-            AccessModifier.privateAccess);
+            AccessModifier.PRIVATE);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -150,7 +150,7 @@ public class ExclusionFilesTest {
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
             new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
-            AccessModifier.privateAccess);
+            AccessModifier.PRIVATE);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -170,7 +170,7 @@ public class ExclusionFilesTest {
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.cloud.google.Foo"),
             new MethodSymbol("com.google.cloud.Foo", "methodA", "()Ljava.lang.String;", false),
-            AccessModifier.privateAccess);
+            AccessModifier.PRIVATE);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -190,7 +190,7 @@ public class ExclusionFilesTest {
             sourceClass,
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.googler.Foo"),
             new MethodSymbol("com.googler.Foo", "methodA", "()Ljava.lang.String;", false),
-            AccessModifier.privateAccess);
+            AccessModifier.PRIVATE);
     boolean result = matcher.match(linkageProblemToMatch);
     assertFalse(result);
   }
@@ -210,7 +210,7 @@ public class ExclusionFilesTest {
                 new ClassPathEntry(Paths.get("dummy.jar")), "reactor.core.publisher.Traces"),
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
             new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
-            AccessModifier.privateAccess);
+            AccessModifier.PRIVATE);
     boolean result = matcher.match(linkageProblemToMatch);
     assertTrue(result);
   }
@@ -229,7 +229,7 @@ public class ExclusionFilesTest {
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Bar"),
             new ClassFile(new ClassPathEntry(Paths.get("dummy.jar")), "com.google.Foo"),
             new MethodSymbol("com.google.Foo", "methodA", "()Ljava.lang.String;", false),
-            AccessModifier.privateAccess);
+            AccessModifier.PRIVATE);
     boolean result = matcher.match(linkageProblemToMatch); // No match
     assertFalse(result);
   }

--- a/enforcer-rules/src/it/inaccessible-class-error/invoker.properties
+++ b/enforcer-rules/src/it/inaccessible-class-error/invoker.properties
@@ -1,0 +1,2 @@
+invoker.goals=verify
+invoker.buildResult=failure

--- a/enforcer-rules/src/it/inaccessible-class-error/pom.xml
+++ b/enforcer-rules/src/it/inaccessible-class-error/pom.xml
@@ -1,0 +1,82 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+  ~ Copyright 2020 Google LLC.
+  ~
+  ~ Licensed under the Apache License, Version 2.0 (the "License");
+  ~ you may not use this file except in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~     http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.google.cloud.tools.opensource</groupId>
+  <artifactId>inaccessible-class-error-example</artifactId>
+  <version>1.0-SNAPSHOT</version>
+
+  <name>inaccessible-class-error-example</name>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.compiler.source>1.8</maven.compiler.source>
+    <maven.compiler.target>1.8</maven.compiler.target>
+  </properties>
+
+  <dependencies>
+    <!-- gax-grpc:1:38's InstantiatingGrpcChannelProvider does not have needsCredentials method of
+         gax:1.48's TransportChannelProvider -->
+    <dependency>
+      <groupId>io.grpc</groupId>
+      <artifactId>grpc-core</artifactId>
+      <version>0.15.0</version>
+    </dependency>
+    <dependency>
+      <groupId>io.grpc</groupId>
+      <artifactId>grpc-grpclb</artifactId>
+      <version>0.12.0</version>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-enforcer-plugin</artifactId>
+        <version>3.0.0-M3</version>
+        <dependencies>
+          <dependency>
+            <groupId>com.google.cloud.tools</groupId>
+            <artifactId>linkage-checker-enforcer-rules</artifactId>
+            <version>@project.version@</version>
+          </dependency>
+        </dependencies>
+        <executions>
+          <execution>
+            <id>enforce-linkage-checker</id>
+            <phase>verify</phase>
+            <goals>
+              <goal>enforce</goal>
+            </goals>
+            <configuration>
+              <rules>
+                <LinkageCheckerRule
+                  implementation="com.google.cloud.tools.dependencies.enforcer.LinkageCheckerRule">
+                </LinkageCheckerRule>
+              </rules>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/enforcer-rules/src/it/inaccessible-class-error/verify.groovy
+++ b/enforcer-rules/src/it/inaccessible-class-error/verify.groovy
@@ -1,0 +1,6 @@
+def buildLog = new File(basedir, "build.log").text.replaceAll("\\r\\n", "\n")
+
+assert buildLog.contains('''\
+(io.grpc:grpc-core:0.15.0) Class io.grpc.internal.SingleTransportChannel has default access;
+  referenced by 1 class file in a different package
+    io.grpc.grpclb.GrpclbLoadBalancer (io.grpc:grpc-grpclb:0.12.0)''')

